### ISSUE TITLE
Missing method to initialize a `DateTime` with non-zero millis or nanos

### ISSUE THREAD

**Issue by Guara92** [chronotope/chrono#873]
Trying to upgrade to 0.4.23 I'm facing a limitation of the trait `TimeZone` where now the only way to get a DateTime is via `with_ymd_and_hms` that assume nanos always 0.
I think that the trait should expose a method to override this behaviour or `ymd_opt` shouldn't be deprecated, in particular my usecase is something like:

```rust
Utc.ymd_opt(2022, 11, 14).unwrap().and_hms_milli_opt(23, 59, 59, 999).unwrap(),
```
I'm missing a way to do this?

**Comment by djc** [comment]
Yes, the recommended way to do this is something like `NaiveDateTime::from_ymd_opt(2022, 11, 14).and_then(|d| d.and_hms_milli_opt(23, 59, 59, 999)).unwrap().and_local_timezone(Utc)`.

**Comment by Guara92** [comment]
Thanks your solution is ok but maybe too verbose? there is a specific reason to deprecate `ymd_opt`?
(the deprecation message is saying `use `with_ymd_and_hms()` instead`)

**Comment by esheppa** [comment]
> there is a specific reason to deprecate ymd_opt?

`ymd_opt` was a method that produced a `Date<Tz>` which has also been deprecated. 

```rust
NaiveDateTime::from_ymd_opt(2022, 11, 14)?.and_hms_milli_opt(23, 59, 59, 999)?.and_local_timezone(Utc)?

```
is only slightly more verbose than

```rust
Utc.ymd_opt(2022, 11, 14)?.and_hms_milli_opt(23, 59, 59, 999)?
```

but has an added advantage that you don't need to import the `TimeZone` trait anymore.

However I'm keen to understand occurance of usages like this in code bases to see where we can possibly improve the API.

From my own experience in codebases, I'm thinking of the following:

* `and_time` is infallible so having a `NaiveTime::START` or `NaiveTime::MIN` would be useful to combine with this, or perhaps even a `NaiveDate::start_time` method. `NaiveTime::END` is problematic because of leap seconds, but the start seems more generally useful
* I often want the first or last day in a given year or month, some thoughts here are `NaiveDate::start_of_year(year: i32)`, `NaiveDate::start_of_month(year: i32, month: Month)` which may be able to be implemented infallibly.

Would these and/or similar methods assist in reducing the total amount of verbosity?

**Comment by djc** [comment]
(We wouldn't want to implement `start_of_year(year: i32, month: Month)` infallibly because the `year` range could still be off.)

**Comment by esheppa** [comment]
I'm trying out some ideas extending https://github.com/chronotope/chrono/pull/845 where the year part of the date is infallible, but represented by a i16. IMO reducing the year range, but having it be infallible could be an excellent tradeoff if it opens up the possibilities mentioned above

**Comment by mathstuf** [comment]
I think this would be a lot easier if something like `Utc.with_ymd_and_hms(…).unwrap().add_millis_opt(…).unwrap()` were possible.

**Comment by djc** [comment]
> I'm trying out some ideas extending #845 where the year part of the date is infallible, but represented by a i16. IMO reducing the year range, but having it be infallible could be an excellent tradeoff if it opens up the possibilities mentioned above

Yeah, that tradeoff definitely makes a lot of sense!

> I think this would be a lot easier if something like `Utc.with_ymd_and_hms(…).unwrap().add_millis_opt(…).unwrap()` were possible.

Anything that has more than one `unwrap()` is suboptimal, of course. Would like to find other ways of doing it.

**Comment by eilidh-spire** [comment]
Came up against this recently, and there isn't really a pretty way around it. In the end I used `Utc.with_ymd_and_hms_opt(...).unwrap() + Duration::milliseconds(...)` 🙃 

Would be really nice to be able to just specify all components at once. How about somthing like `Utc.with_ymd_and_hms_and_millis_opt(...).unwrap()`? Too unwieldy?

**Comment by esheppa** [comment]
@eilidh-spire  do any of the solutions proposed in #815 look ok? These could be extended as you describe with a milliseconds parameter as well
